badge_number.go

```diff
@@ -0,0 +1,53 @@
+package apns
+
+import "encoding/json"
+
+// BadgeNumber is much a NullInt64 in
+// database/sql except instead of using
+// the nullable value for driver.Value
+// encoding and decoding, this is specifically
+// meant for JSON encoding and decoding
+type BadgeNumber struct {
+	Number uint
+	IsSet  bool
+}
+
+// Unset will reset the BadgeNumber to
+// both 0 and invalid
+func (b *BadgeNumber) Unset() {
+	b.Number = 0
+	b.IsSet = false
+}
+
+// Set will set the BadgeNumber value to
+// the number passed in, assuming it's >= 0.
+// If so, the BadgeNumber will also be marked valid
+func (b *BadgeNumber) Set(number uint) {
+	b.Number = number
+	b.IsSet = true
+}
+
+// MarshalJSON will marshall the numerical value of
+// BadgeNumber
+func (b BadgeNumber) MarshalJSON() ([]byte, error) {
```

Comment:
Shoot, there is one more thing. Do we know why this method takes a copy of the struct, and the rest of the methods take a pointer to the struct? Usually I only see the pointer to the struct, though there are instances where a copy is wanted.